api/src/main/java/jakarta/enterprise/inject/build/compatible/spi/AnnotationBuilder.java

```diff
@@ -67,7 +70,7 @@ default AnnotationBuilder value(boolean value) {
      * @param values the boolean array, must not be {@code null}
      * @return this {@code AnnotationBuilder}
      */
-    default AnnotationBuilder value(boolean... values) {
+    default AnnotationBuilder value(boolean[] values) {
```

Comment:
I'm not 100% sure about this change, but when I tried to use the API, it was actually a bit surprising to me. I started wondering whether we should do something about single-element arrays, maybe automatically convert the type or so. In the end, I thought getting rid of  varargs and forcing explicit arrays is a nice way out of it.

Reply:
I prefer varargs instead of the array. In the previous form, you can pass one value while the proposed change needs to convert the single value to an array.

Reply:
That is in fact incorrect, and shows exactly the same confusion I found myself in.

Note that we have overloaded methods `value(boolean)` and `value(boolean...)`. If you call `value(true)`, the overload resolution process will always select the first method (see https://docs.oracle.com/javase/specs/jls/se17/html/jls-15.html#jls-15.12.2: _This ensures that a method is never chosen through variable arity method invocation if it is applicable through fixed arity method invocation_).

Varargs here make sense if _more than 1_ argument is passed, or in case of an empty array. But for a single-argument array, you would always have to write `new boolean[] { true }`, which is inconsistent and, as demonstrated above, confusing.

Reply:
Another method to eliminate the confusion would be, of course, to eliminate the overloads, so that the varargs accepting method would be called differently. I don't have a good name, personally.

Reply:
> Another method to eliminate the confusion would be, of course, to eliminate the overloads

That's what I was thinking when I saw your comment. Couldn't we have a `value` method name for single value and `values` for varargs? Or something like `singleValue` and `value`. Basically, I'd just look for some names so that we are able to keep varargs variants in place.

Reply:
The name `values` makes sense.

Reply:
The name `values` doesn't make sense, because the method defines a value of the annotation member called `value`.

Having only `value(boolean...)` would be possible, but would require implementations to find what's the type of the annotation member, which necessitates verifying that given annotation member even exists, which is something I'd like to avoid. But maybe there's a way I don't see.

Reply:
To clarify further.

One reason why I don't want to force implementations to find types of annotation members is that I think it should be possible to define values for annotation members that don't exist (those values would be ignored). The reason is hypothetical forward compatibility. It should be possible to define a value for an annotation member that doesn't exist on an older version of the annotation, but does exist on a newer version of the annotation.

Maybe that's not entirely useful and we should fail in such situation -- I can see benefits in both ways and don't really mind. (I just pushed a change where I explicitly specify that nonexisting members are ignored. I can amend it again to say the opposite.)

Another reason is that I think callers should be _explicit_ about which type of the annotation member they are defining. On other places, we moved away from type coercion, I don't think we should add it here. When I read `value(true)`, I need to know immediately what type is it -- I don't want to look into the annotation definition to figure out if it's singular or array. Principle of locality.

Reply:
personally I think we should have `value(boolean)` and `value(boolean[])` it is not worth the hassle to support varargs and yes we 100% need the ability to specify meta-members that don't actually exist

Reply:
Just out of curiosity @graemerocher, what is your reason to support specifying annotation members that don't exist? I admit mine is just the hypothetical forward compatibility -- I'd love to know if you have more.

Reply:
We use it in Micronaut Data. We precompute the SQL/JPA-QL query and store it a meta annotation member that has actual real equivalent

Reply:
That's an interesting one, thanks for sharing!